warehouse/bigquery.go

```diff
@@ -0,0 +1,394 @@
+
+	q := fmt.Sprintf("SELECT COUNT(*) FROM %s.__TABLES_SUMMARY__ WHERE table_id = '%s';", bq.conf.BigQuery.Dataset, name)
+	query := bq.bqClient.Query(q)
+	query.QueryConfig.UseStandardSQL = true
+
+	iter, err := query.Read(bq.ctx)
+	if err != nil {
+		// something is horribly wrong...just give up
+		log.Fatal(err)
+	}
+
+	var row []bigquery.Value
+	err = iter.Next(&row)
+	if err != nil {
+		// not checking for iterator.Done, as count queries should always return at least 1 row
+		log.Fatal(err)
+	}
+
+	cnt := row[0].(int64)
+	return cnt > 0
+}
+
+func (bq *BigQuery) createSyncTable() error {
+	log.Printf("Creating table %s", bq.conf.BigQuery.SyncTable)
+
+	schema, err := bigquery.InferSchema(syncTable{})
+	if err != nil {
+		return err
+	}
+
+	table := bq.bqClient.Dataset(bq.conf.BigQuery.Dataset).Table(bq.conf.BigQuery.SyncTable)
+	if err := table.Create(bq.ctx, schema); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bq *BigQuery) createExportTable() error {
+	log.Printf("Creating table %s", bq.conf.BigQuery.ExportTable)
+
+	schema, err := bigquery.InferSchema(exportSchema{})
+	if err != nil {
+		return err
+	}
+
+	// only EventStart and EventType should be required
+	for i := 2; i < len(schema); i++ {
+		schema[i].Required = false
+	}
+
+	table := bq.bqClient.Dataset(bq.conf.BigQuery.Dataset).Table(bq.conf.BigQuery.ExportTable)
+	// create export table as date partitioned, with no expiration date (it can be set later)
+	if err := table.Create(bq.ctx, schema, &bigquery.TimePartitioning{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bq *BigQuery) fetchTimeVal(q string, time *pq.NullTime) error {
+	query := bq.bqClient.Query(q)
+	query.QueryConfig.UseStandardSQL = true
+
+	iter, err := query.Read(bq.ctx)
+	if err != nil {
+		log.Printf("Could not run query %q because: %s", q, err)
+		time.Valid = false
+		return err
+	}
+
+	var row []bigquery.Value
+	err = iter.Next(&row)
+	if err != nil {
+		log.Printf("Could not fetch result for query %q because: %s", q, err)
+		time.Valid = false
+		return err
+	}
+
+	time.Scan(row[0])
+	return nil
+}
+
+func (bq *BigQuery) latestEventStart(time *pq.NullTime) error {
+	if !bq.doesTableExist(bq.conf.BigQuery.ExportTable) {
+		time.Valid = false
+		return nil
+	}
+
+	// export table exists, get latest EventStart from it
+	q := fmt.Sprintf("SELECT max(EventStart) FROM %s.%s;", bq.conf.BigQuery.Dataset, bq.conf.BigQuery.ExportTable)
+	return bq.fetchTimeVal(q, time)
+}
+
+func (bq *BigQuery) uploadToGCS(filename string) (string, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	gcsClient, err := storage.NewClient(bq.ctx)
+	if err != nil {
+		return "", err
+	}
+	defer gcsClient.Close()
+
+	_, objName := filepath.Split(filename)
+	w := gcsClient.Bucket(bq.conf.GCS.Bucket).Object(objName).NewWriter(bq.ctx)
+	if _, err = io.Copy(w, f); err != nil {
+		return "", err
+	}
+	if err := w.Close(); err != nil {
```

Comment:
I'm not very familiar with the GCS api. Should this be a `defer w.Close()` right after the NewWriter is created?

Reply:
It does seem inconsistent with how you would normally handle closing something, but all the samples are written this way, both for creating (https://cloud.google.com/appengine/docs/standard/go/googlecloudstorageclient/app-engine-cloud-storage-sample) and uploading files to GCS (https://github.com/GoogleCloudPlatform/golang-samples/blob/master/storage/gcsupload/gcsupload.go).  

While googling for more about this, I did find one guy who argued that you shouldn't ignore the err resulting from closing a file Writer, so that you can be 100% all your data really did get written to the file (https://joeshaw.org/dont-defer-close-on-writable-files/).  I would imagine the GCS samples are written this way for reasons that are similar in pattern, if not in exact origin.